src/averageAges.js

```diff
@@ -55,10 +76,25 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const peopleWhitMother = people.filter(person => !onlyWithSon
+    ? people.find(women => person.mother === women.name)
+    : (people.find(women => person.mother === women.name
+      && person.sex === MALE)
+    ));
+
+  const womenMother = people.filter(mother =>
+    peopleWhitMother.some(child => mother.name === child.mother));
+
+  const sumOfDifference = peopleWhitMother
+    .map(child => {
+      const motherOfPerson = womenMother
+        .find(woman => woman.name === child.mother);
+
+      return child.born - motherOfPerson.born;
+    })
+    .reduce((acc, value) => acc + value, 0);
```

Comment:
Move the calculation from the `map` inside the `reduce`, so as not to go through the array twice.